addons/mail/models/mail_presence.py

```diff
@@ -11,33 +11,41 @@
 PRESENCE_OUTDATED_TIMER = 12 * 60 * 60  # 12 hours
 
 
-class BusPresence(models.Model):
+class MailPresence(models.Model):
     """ User Presence
         Its status is 'online', 'away' or 'offline'. This model should be a one2one, but is not
         attached to res_users to avoid database concurrence errors. Since the 'update_presence' method is executed
         at each poll, if the user have multiple opened tabs, concurrence errors can happend, but are 'muted-logged'.
     """
 
+    _name = "mail.presence"
     _description = 'User Presence'
     _log_access = False
 
     user_id = fields.Many2one('res.users', 'Users', ondelete='cascade')
+    guest_id = fields.Many2one('mail.guest', 'Guest', ondelete='cascade')
     last_poll = fields.Datetime('Last Poll', default=lambda self: fields.Datetime.now())
     last_presence = fields.Datetime('Last Presence', default=lambda self: fields.Datetime.now())
     status = fields.Selection([('online', 'Online'), ('away', 'Away'), ('offline', 'Offline')], 'IM Status', default='offline')
 
+    _guest_unique = models.UniqueIndex("(guest_id) WHERE guest_id IS NOT NULL")
     _user_unique = models.UniqueIndex("(user_id) WHERE user_id IS NOT NULL")
 
+    _partner_or_guest_exists = models.Constraint(
+        "CHECK((user_id IS NOT NULL AND guest_id IS NULL) OR (user_id IS NULL AND guest_id IS NOT NULL))",
+        "A mail presence must have a user or a guest."
+    )
+
     def create(self, values):
         presences = super().create(values)
         presences._invalidate_im_status()
         presences._send_presence()
         return presences
 
     def write(self, values):
-        status_by_user = {presence._get_identity_field_name(): presence.status for presence in self}
+        status_by_user = {"guest_id" if presence.guest_id else "user_id": presence.status for presence in self}
         result = super().write(values)
-        updated = self.filtered(lambda p: status_by_user[p._get_identity_field_name()] != p.status)
+        updated = self.filtered(lambda p: status_by_user["guest_id" if p.guest_id else "user_id"] != p.status)
```

Comment:
Now that I think about it, we don't need a "status by user", we can just have a "status by presence" which requires no condition on user or guest.